src/libraries/System.Security.Cryptography.Encoding/tests/Cbor/CborReader.Tag.cs

```diff
@@ -0,0 +1,157 @@
+        }
+
+        public CborTag PeekTag() => PeekTagCore(out int _);
+
+        private CborTag PeekTagCore(out int additionalBytes)
+        {
+            CborInitialByte header = PeekInitialByte(expectedType: CborMajorType.Tag);
+            return (CborTag)ReadUnsignedInteger(_buffer.Span, header, out additionalBytes);
+        }
+
+        // Additional tagged type support
+
+        public DateTimeOffset ReadDateTimeOffset()
+        {
+            // implements https://tools.ietf.org/html/rfc7049#section-2.4.1
+
+            switch (PeekTag())
+            {
+                case CborTag.DateTimeString:
+                    ReadTag();
+
+                    if (Peek() != CborReaderState.TextString)
+                    {
+                        throw new FormatException("String DateTime semantic tag should annotate string value.");
+                    }
+
+                    string dateString = ReadTextString();
+
+                    if (!DateTimeOffset.TryParseExact(dateString, CborWriter.Rfc3339FormatString, null, DateTimeStyles.RoundtripKind, out DateTimeOffset result))
+                    {
+                        throw new FormatException("DateTime string is not valid RFC3339.");
+                    }
+
+                    return result;
+
+                case CborTag.DateTimeUnixSeconds:
+                    ReadTag();
+
+                    switch (Peek())
+                    {
+                        case CborReaderState.UnsignedInteger:
+                        case CborReaderState.NegativeInteger:
+                            return DateTimeOffset.FromUnixTimeSeconds(ReadInt64());
+
+                        case CborReaderState.HalfPrecisionFloat:
+                        case CborReaderState.SinglePrecisionFloat:
+                        case CborReaderState.DoublePrecisionFloat:
+                            // we don't (but probably should) have a float overload for DateTimeOffset.FromUnixTimeSeconds
+                            double seconds = ReadDouble();
+                            long epochTicks = DateTimeOffset.UnixEpoch.Ticks;
+                            long ticks = checked(epochTicks + (long)(seconds * TimeSpan.TicksPerSecond));
+                            return new DateTimeOffset(ticks, TimeSpan.Zero);
+
+                        default:
+                            throw new FormatException("Epoch DateTime semantic tag should annotate numeric value.");
+                    }
+
+                default:
+                    throw new InvalidOperationException("CBOR tag is not a recognized DateTime value.");
+            }
+        }
+
+        public BigInteger ReadBigInteger()
+        {
+            bool isUnsigned = PeekTag() switch
+            {
+                CborTag.UnsignedBigNum => true,
+                CborTag.NegativeBigNum => false,
+                _ => throw new InvalidOperationException("CBOR tag is not a recognized Bignum value."),
+            };
+
+            ReadTag();
```

Comment:
I can go either way here, but I have to ask:

Since the DateTimeOffset method supports multiple formats, should ReadBigInteger also support major type 0 and major type 1, since they can be upconverted losslessly?

I haven't looked at a whole lot of schema, so I don't know if any are flexible on "small integer" vs "big integer" encodings.  The only one I can really think of that could even really theoretically use numbers in that range is RSA public key encoding; but https://tools.ietf.org/html/rfc8230#section-4 defines its own encoding (Unsigned Big-Endian value with extraneous leading zeros removed encoded into a byte string).

And, of course, someone could add support for it with an extension method, like

```C#
public static ReadAnyBigInteger(this CborReader reader)
{
    switch (reader.State)
    {
        case StateEnum.NonNegativeInteger:
        case StateEnum.NegativeInteger:
            return reader.ReadInt64();
        case StateEnum.TaggedValue:
            return reader.ReadBigInteger();
    }

    throw new Exception("I'm too lazy to write good messages in hypothetical code");
}
```

Reply:
> Since the DateTimeOffset method supports multiple formats, should ReadBigInteger also support major type 0 and major type 1, since they can be upconverted losslessly?

I was working under the assumption that we were avoiding implicit conversions by design. For the case of `ReadDateTimeOffset()` I'm contemplating whether I should revert to distinct readers for each tagged type.